main.go

```diff
@@ -199,16 +204,52 @@ func StartPurgeLoop(secInterval int) {
 func writeToPumps(keys []interface{}, job *health.Job, startTime time.Time) {
 	// Send to pumps
 	if Pumps != nil {
+		var wg sync.WaitGroup
+		wg.Add(len(Pumps))
+
 		for _, pmp := range Pumps {
-			log.WithFields(logrus.Fields{
-				"prefix": mainPrefix,
-			}).Debug("Writing to: ", pmp.GetName())
-			pmp.WriteData(keys)
-			if job != nil {
-				job.Timing("purge_time_"+pmp.GetName(), time.Since(startTime).Nanoseconds())
-			}
+			go func(wg *sync.WaitGroup, pmp pumps.Pump) {
+				defer wg.Done()
+				ch := make(chan error, 1)
+
+				//Load pump timeout
+				timeout, hasTimeout := PumpsTimeout.Load(pmp.GetName())
+				var ctx context.Context
+				var cancel context.CancelFunc
+				//Initialize context depending if the pump has a configured timeout
+				if hasTimeout && timeout.(int) > 0 {
+					ctx, cancel = context.WithTimeout(context.Background(), time.Duration(timeout.(int))*time.Second)
+				} else {
+					ctx, cancel = context.WithCancel(context.Background())
+				}
+
+				defer cancel()
+
+				go func(ch chan error, ctx context.Context, pmp pumps.Pump, keys []interface{}) {
+					err := pmp.WriteData(ctx, keys)
+					ch <- err
+				}(ch, ctx, pmp, keys)
+
+				select {
+				case err := <-ch:
+					if err != nil {
+						log.WithFields(logrus.Fields{
+							"prefix": mainPrefix,
+						}).Warning("Error Writing to: ", pmp.GetName(), " - Error:", err)
+					}
+				case <-ctx.Done():
+					log.WithFields(logrus.Fields{
+						"prefix": mainPrefix,
+					}).Warning("Timeout Writing to: ", pmp.GetName())
+				}
+
+				if job != nil {
+					job.Timing("purge_time_"+pmp.GetName(), time.Since(startTime).Nanoseconds())
+				}
 
+			}(&wg, pmp)
 		}
+		wg.Wait()
```

Comment:
We need a way to tell the user that he having an issue, even if timeout is not set.
We need to print a warning that if pump processing takes more time then purge_delay, and propose to set timeout option if needed.

Reply:
Added 2 warning messages for each pump:
**If pump processing is taking more time than purge_delay and no timeout is configure:**
Pump  _pump_name_ is taking more time than the value configured of purge_delay. You should try to set a timeout for this pump. 
**if pump processing is taking more time than purge_delay and timeout > purge_delay:**
Pump  _pump_name_ is taking more time than the value configured of purge_delay. You should try lowering the timeout configured for this pump.

Do you think it’s okay? @buger